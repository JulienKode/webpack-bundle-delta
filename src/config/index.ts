/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { cosmiconfig } from 'cosmiconfig';
import allPluginsConfig from './allPlugins';
import {
  BasePluginConfig,
  defaultBasePluginConfig,
  PluginConfig,
  UserDefinedConfig,
} from './PluginConfig';

const moduleName = 'webpackBundleDelta';

const getUserConfig = async (): Promise<UserDefinedConfig> => {
  const searchedFor = await cosmiconfig(moduleName).search();
  return {
    ...(searchedFor && !searchedFor.isEmpty ? searchedFor.config : { filepath: '' }),
  };
};

// we don't want to override is `extends: null` for example
const useInternalDefaults = (userConfig: UserDefinedConfig) =>
  !Object.keys(userConfig).includes('extends');

const applyGlobals = (
  globals: BasePluginConfig,
  plugins: PluginConfig[],
  globalsOverrides = false
) =>
  (plugins || []).map((plugin) => {
    if (typeof plugin.config === 'object') {
      if (globalsOverrides) {
        // eslint-disable-next-line no-param-reassign
        plugin.config = {
          ...plugin.config,
          ...globals,
        };
      } else {
        // eslint-disable-next-line no-param-reassign
        plugin.config = {
          ...globals,
          ...plugin.config,
        };
      }
    } else if (typeof plugin.config === 'boolean' && plugin.config) {
      // eslint-disable-next-line no-param-reassign
      plugin.config = { ...defaultBasePluginConfig, ...globals };
    }

    return plugin;
  });

const mergePlugins = (
  basePlugins: PluginConfig[],
  extendingPlugins: PluginConfig[]
): PluginConfig[] => {
  const pluginMap = [basePlugins, extendingPlugins].reduce((mergedPlugins, plugins) => {
    plugins?.forEach(({ name, config }) => {
      let mergedConfig = mergedPlugins[name];
      if (typeof mergedConfig === 'object' && typeof config === 'object') {
        mergedConfig = { ...(mergedPlugins[name] as Record<string, unknown>), ...config };
      } else if (typeof config === 'object') {
        mergedConfig = { ...defaultBasePluginConfig, ...config };
      } else if (!config) {
        mergedConfig = false;
      } else {
        mergedConfig = defaultBasePluginConfig;
      }
      // eslint-disable-next-line no-param-reassign
      mergedPlugins[name] = mergedConfig;
    });
    return mergedPlugins;
  }, {} as { [pluginName: string]: unknown });

  return Object.entries(pluginMap).map(([name, config]) => ({ name, config } as PluginConfig));
};

const extendConfig = async (extendsPath: string): Promise<UserDefinedConfig> => {
  const [path, section] = extendsPath.split(':');
  let baseConfig: UserDefinedConfig = {};
  // eslint-disable-next-line global-require, import/no-dynamic-require, @typescript-eslint/no-var-requires
  let extendingConfig: UserDefinedConfig = require(path);
  if (section) {
    extendingConfig = (extendingConfig as { [key: string]: UserDefinedConfig })[section];
  }
  if (!extendingConfig) {
    throw new Error(
      `Unable to find section "${section}" in config "${path}": is it a named export (i.e. module.exports = { ${section}: someConfig };)?`
    );
  }
  if (extendingConfig.extends) {
    baseConfig = await extendConfig(extendingConfig.extends);
  } else if (useInternalDefaults(extendingConfig) && !extendingConfig.plugins?.length) {
    baseConfig.plugins = applyGlobals(extendingConfig.globals, allPluginsConfig, true);
  }

  if (extendingConfig.globals) {
    baseConfig.plugins = applyGlobals(extendingConfig.globals, baseConfig.plugins, true);
  }

  return {
    extends: extendsPath,
    plugins: mergePlugins(
      applyGlobals(baseConfig.globals, baseConfig.plugins),
      applyGlobals(extendingConfig.globals, extendingConfig.plugins)
    ),
  };
};

const createConfig = async (): Promise<UserDefinedConfig> => {
  const userConfig = await getUserConfig();
  let baseConfig: UserDefinedConfig = {};

  if (userConfig.extends) {
    baseConfig = await extendConfig(userConfig.extends);
  } else if (useInternalDefaults(userConfig)) {
    baseConfig = {
      globals: defaultBasePluginConfig,
      plugins: applyGlobals(userConfig.globals, allPluginsConfig, true),
    };
  }

  return {
    ...userConfig,
    plugins: mergePlugins(
      applyGlobals(baseConfig.globals, baseConfig.plugins),
      applyGlobals(userConfig.globals, userConfig.plugins)
    ),
  };
};

export default createConfig;
