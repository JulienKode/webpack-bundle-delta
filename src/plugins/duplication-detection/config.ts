/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { BasePluginConfig, defaultBasePluginConfig } from '../../config/PluginConfig';

export interface DuplicationDetectionConfig extends BasePluginConfig {
  /**
   * Minimum size of the module for duplication detection
   *
   * Value in format of "<number><measurement>"
   * If size of the whole module (i.e. sum of all individual files within the module) is less than specified value
   * It will not be counted into the list of duplicates
   *
   * @type {string}
   * @memberof Duplication
   * @example
   * "150KB"
   * "1.52MB"
   */
  minSize: string;
  /**
   * List of modules to ignore
   *
   * Either module name or package sub module path
   * Ignoring sub modules may be necessary if a module is using a different version that isn't compatible with other versions within the application
   * This can occur when the module isn't upgraded to the newer version
   *
   * @type {string[]}
   * @memberof Duplication
   * "lodash"
   * "some-module/node_modules/lodash"
   */
  ignore: string[];
}

export const defaultDuplicationDetectionConfig: DuplicationDetectionConfig = {
  ...defaultBasePluginConfig,
  minSize: '5KB',
  ignore: null,
};
