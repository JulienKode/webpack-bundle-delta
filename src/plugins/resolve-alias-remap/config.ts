/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { BasePluginConfig, defaultBasePluginConfig } from '../../config/PluginConfig';
import { AliasRemap } from './AliasRemap';
import lodashSuggestion from './suggestions/lodash';

/**
 * Resolve alias remap configuration
 *
 * @export
 * @interface ResolveAliasRemapConfig
 */
export interface ResolveAliasRemapConfig extends BasePluginConfig {
  /**
   * Keep default remap
   *
   * By default, when configurations are specified, they'll override the default values
   * Specifying this flag ensures the default remap is kept whilst the ones specified are also used
   *
   * @type {boolean}
   * @memberof ResolveAliasRemapConfig
   */
  keepDefaultRemap: boolean;
  /**
   * An array of remap logic used to determine suggestions
   *
   * @type {AliasRemap[]}
   * @memberof ResolveAliasRemapConfig
   */
  remap: AliasRemap[];
}

export const defaultResolveAliasRemapConfig: ResolveAliasRemapConfig = {
  ...defaultBasePluginConfig,
  keepDefaultRemap: false,
  remap: [lodashSuggestion],
};
