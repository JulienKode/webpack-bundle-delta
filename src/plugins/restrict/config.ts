/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */

import { BasePluginConfig } from '../../config/PluginConfig';

export interface Restriction {
  /**
   * What to search for in the given module name
   *
   * The value will be passed to `RegExp` for verification
   *
   * @type {(string | RegExp)}
   * @memberof Restriction
   */
  search: string | RegExp;
  /**
   * Action to take
   *
   * @type {('info' | 'warn' | 'error')}
   * @memberof Restriction
   */
  responseType: 'info' | 'warn' | 'error';
  /**
   * Message to display when matched
   *
   * @type {string}
   * @memberof Restriction
   */
  message: string;
}

/**
 * Restrict configuration
 *
 * @export
 * @interface RestrictConfig
 */
export interface RestrictConfig extends BasePluginConfig {
  /**
   * Array of restrictions to be applied
   *
   * @type {Restriction[]}
   * @memberof RestrictConfig
   */
  restrictions: Restriction[];
}

/** default is set to false as we don't want to enable this plugin by default */
export const defaultRestrictConfig: RestrictConfig | boolean = false;
