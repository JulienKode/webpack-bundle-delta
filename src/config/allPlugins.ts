/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { PluginConfig } from './PluginConfig';
import { defaultSizeChangesConfig } from '../plugins/size-changes/config';
import { defaultDuplicationDetectionConfig } from '../plugins/duplication-detection/config';
import { defaultResolveAliasRemapConfig } from '../plugins/resolve-alias-remap/config';
import { defaultRestrictConfig } from '../plugins/restrict/config';

const allPlugins: PluginConfig[] = [
  {
    name: 'size-changes',
    config: defaultSizeChangesConfig,
  },
  {
    name: 'trace-changes',
    config: true,
  },
  {
    name: 'duplication-detection',
    config: defaultDuplicationDetectionConfig,
  },
  {
    name: 'resolve-alias-remap',
    config: defaultResolveAliasRemapConfig,
  },
  {
    name: 'restrict',
    config: defaultRestrictConfig,
  },
];

export default allPlugins;
