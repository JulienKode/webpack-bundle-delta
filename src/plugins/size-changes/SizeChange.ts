/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import { BudgetConfig } from './config';

export interface SizeChange {
  file: string;
  prevSize?: number;
  prevGzSize?: number;
  prevBrSize?: number;
  currentSize?: number;
  currentGzSize?: number;
  currentBrSize?: number;
  sizeDiff?: number;
  gzSizeDiff?: number;
  brSizeDiff?: number;
  significantFields: SignificantField[];
  budget?: BudgetResult;
}

export enum BudgetResultType {
  'size' = 'size',
  'gzSize' = 'gzSize',
  'brSize' = 'brSize',
}

export interface BudgetResult {
  amount: number;
  type: BudgetResultType;
  matched: BudgetConfig;
}

export enum SignificantField {
  sizeDiffBytes = 'sizeDiffBytes',
  sizeDiffPercent = 'sizeDiffPercent',
  gzSizeDiffBytes = 'gzSizeDiffBytes',
  gzSizeDiffPercent = 'gzSizeDiffPercent',
  brSizeDiffBytes = 'brSizeDiffBytes',
  brSizeDiffPercent = 'brSizeDiffPercent',
}
