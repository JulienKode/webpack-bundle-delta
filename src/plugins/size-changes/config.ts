/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */

import { BasePluginConfig, defaultBasePluginConfig } from '../../config/PluginConfig';

export interface SignificanceThreshold {
  /**
   * Stat size difference in bytes
   * Number > 0
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  sizeDiffBytes?: number;
  /**
   * Stat size difference as a percentage
   * Number > 0
   * example: 15 means 15%
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  sizeDiffPercent?: number;
  /**
   * Gzip size difference in bytes
   * Number > 0
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  gzSizeDiffBytes?: number;
  /**
   * Gzip size difference as a percentage
   * Number > 0
   * example: 15 means 15%
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  gzSizeDiffPercent?: number;
  /**
   * Brotli size difference in bytes
   * Number > 0
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  brSizeDiffBytes?: number;
  /**
   * Brotli size difference as a percentage
   * Number > 0
   * example: 15 means 15%
   *
   * @type {number}
   * @memberof SignificanceThreshold
   */
  brSizeDiffPercent?: number;
}

/**
 * An array of budget configurations
 *
 * @export
 * @interface Budget
 * @extends {Array<BudgetConfig>}
 */
export type Budget = Array<BudgetConfig>;

/**
 * Budget configuration
 *
 * At least 1 size option is required for budget comparison
 *
 * If the budget is exceeded:
 * - In CLI: error message will be outputted and exit code > 0
 * - In DangerJS: error message will be reported
 *
 * @export
 * @interface BudgetConfig
 */
export interface BudgetConfig {
  /**
   * Chunk name to compute against.
   *
   * String is either the chunk name or a regular expression string
   *
   * @type {string}
   * @memberof BudgetConfig
   * @example
   * "App.mjs"
   * "chunk-group-(.+).js"
   */
  chunkName: string;
  /**
   * Stat size budget
   *
   * Value in format of "<number><measurement>"
   *
   * @type {string}
   * @memberof BudgetConfig
   * @example
   * "150KB"
   * "1.52MB"
   */
  size?: string;
  /**
   * Gzip size budget
   *
   * Value in format of "<number><measurement>"
   *
   * @type {string}
   * @memberof BudgetConfig
   * @example
   * "150KB"
   * "1.52MB"
   */
  gzSize?: string;
  /**
   * Brotli size budget
   *
   * Value in format of "<number><measurement>"
   *
   * @type {string}
   * @memberof BudgetConfig
   * @example
   * "150KB"
   * "1.52MB"
   */
  brSize?: string;
  /**
   * Warning threshold as a percentage value (0 to 100)
   *
   * When specified and computed to be a warning (and not over specified size):
   * - in CLI: warning message will be produced, but exit code will remain 0
   * - in DangerJS: warning message will be reported
   *
   * @type {number}
   * @memberof BudgetConfig
   * @example
   * // 10 percent would be
   * 10
   * // 86 percent would be
   * 86
   */
  warnPercentage?: number;
}

export interface SizeChangesConfig extends BasePluginConfig {
  significance: SignificanceThreshold;
  budget?: Budget;
  hideMinorChanges?: boolean;
}

export const defaultSizeChangesConfig: SizeChangesConfig = {
  ...defaultBasePluginConfig,
  significance: {
    sizeDiffBytes: 1000,
    sizeDiffPercent: 5,
    gzSizeDiffPercent: 5,
    brSizeDiffPercent: 5,
  },
  budget: null,
  hideMinorChanges: false,
};
