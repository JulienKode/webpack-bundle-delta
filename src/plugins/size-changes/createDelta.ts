/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import percentageChange from '../../helpers/percentageChange';
import { SignificanceThreshold, Budget } from './config';
import { ChunkNamesExtensionsToSize } from '../../helpers/mapChunkNamesExtensionsToSize';
import significantFields from './significantFields';
import computeBudget from './budget';
import { SizeChange } from './SizeChange';
import { SupportedExtensions } from '../../helpers/constants';

export interface CreateDeltaOptions {
  baseChunkSizes: ChunkNamesExtensionsToSize;
  headChunkSizes: ChunkNamesExtensionsToSize;
  significance: SignificanceThreshold;
  budget: Budget;
}

interface ExtendEntryDetailsParameters {
  entry: SizeChange;
  significance: SignificanceThreshold;
  budget: Budget;
}

const extendEntryDetails = ({
  entry,
  significance,
  budget,
}: ExtendEntryDetailsParameters): SizeChange => {
  return {
    ...entry,
    significantFields: significantFields(entry, significance),
    budget: computeBudget(entry, budget),
  };
};

const createDelta = ({
  baseChunkSizes,
  headChunkSizes,
  significance,
  budget,
}: CreateDeltaOptions): SizeChange[] => {
  const delta: SizeChange[] = [];
  const newEntries: SizeChange[] = [];
  const removedEntries: SizeChange[] = [];

  Object.entries(baseChunkSizes).forEach(([name, data]) => {
    Object.entries(data).forEach(([extension, { size, gzSize, brSize }]) => {
      let entry: SizeChange = {
        file: `${name}.${extension}`,
        prevSize: size,
        prevGzSize: gzSize,
        prevBrSize: brSize,
        currentSize: null,
        currentGzSize: null,
        currentBrSize: null,
        sizeDiff: -100,
        gzSizeDiff: -100,
        brSizeDiff: -100,
        significantFields: [],
      };
      if (
        typeof headChunkSizes[name] !== 'undefined' &&
        typeof headChunkSizes[name][extension as SupportedExtensions] !== 'undefined'
      ) {
        // chunk in both base and head
        const current = headChunkSizes[name][extension as SupportedExtensions];
        const currentSize = current.size;
        const sizeDiff = percentageChange(entry.prevSize, currentSize);
        const gzSizeDiff = percentageChange(entry.prevGzSize, current.gzSize);
        const brSizeDiff = percentageChange(entry.prevBrSize, current.brSize);
        entry = {
          ...entry,
          currentSize: current.size,
          currentGzSize: current.gzSize,
          currentBrSize: current.brSize,
          sizeDiff,
          gzSizeDiff,
          brSizeDiff,
        };
      }
      removedEntries.push(extendEntryDetails({ entry, significance, budget }));
    });
  });

  Object.entries(headChunkSizes).forEach(([name, data]) => {
    Object.entries(data).forEach(([extension, { size, gzSize, brSize }]) => {
      if (
        typeof baseChunkSizes[name] === 'undefined' ||
        typeof baseChunkSizes[name][extension as SupportedExtensions] === 'undefined'
      ) {
        const entry: SizeChange = {
          file: `${name}.${extension}`,
          prevSize: null,
          prevGzSize: null,
          prevBrSize: null,
          currentSize: size,
          currentGzSize: gzSize,
          currentBrSize: brSize,
          sizeDiff: 100,
          gzSizeDiff: 100,
          brSizeDiff: 100,
          significantFields: [],
        };
        newEntries.push(extendEntryDetails({ entry, significance, budget }));
      }
    });
  });

  const entriesBySizeDiff = []
    .concat(newEntries, delta, removedEntries)
    .sort((a: SizeChange, b: SizeChange) => {
      if (Math.abs(a.sizeDiff) > Math.abs(b.sizeDiff)) {
        return -1;
      }
      if (Math.abs(a.sizeDiff) < Math.abs(b.sizeDiff)) {
        return 1;
      }
      if (a.file < b.file) {
        return -1;
      }
      return 1;
    }) as SizeChange[];

  return entriesBySizeDiff;
};

export default createDelta;
