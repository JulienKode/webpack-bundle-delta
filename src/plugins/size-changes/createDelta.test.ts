/*
 * Copyright (c) Trainline Limited, 2020. All rights reserved.
 * See LICENSE.md in the project root for license information.
 */
import createDelta from './createDelta';
import { SizeChange, BudgetResult, BudgetResultType, SignificantField } from './SizeChange';
import { ChunkNamesExtensionsToSize } from '../../helpers/mapChunkNamesExtensionsToSize';
import { defaultSizeChangesConfig } from './config';

const baseChunkSizes: ChunkNamesExtensionsToSize = {
  sgpTrainTimesPage: {
    css: {
      brSize: 37028,
      gzSize: 45562,
      size: 296997,
    },
    js: {
      brSize: 134674,
      gzSize: 174867,
      size: 653190,
    },
    mjs: {
      brSize: 104674,
      gzSize: 144867,
      size: 603190,
    },
  },
};

const headChunkSizes: ChunkNamesExtensionsToSize = {
  sgpTrainTimesPage: {
    css: {
      brSize: 36028,
      gzSize: 42562,
      size: 286997,
    },
    js: {
      brSize: 34674,
      gzSize: 74867,
      size: 553190,
    },
    mjs: {
      brSize: 104674,
      gzSize: 144867,
      size: 603190,
    },
  },
};

describe('createDelta', () => {
  it('returns -100 delta when chunk is in base but not in head', () => {
    const negativeDelta: SizeChange = {
      currentSize: null,
      currentGzSize: null,
      currentBrSize: null,
      prevSize: expect.any(Number),
      prevGzSize: expect.any(Number),
      prevBrSize: expect.any(Number),
      sizeDiff: -100,
      gzSizeDiff: -100,
      brSizeDiff: -100,
      file: expect.any(String),
      significantFields: [
        SignificantField.sizeDiffBytes,
        SignificantField.sizeDiffPercent,
        SignificantField.gzSizeDiffPercent,
        SignificantField.brSizeDiffPercent,
      ],
    };

    expect(
      createDelta({
        baseChunkSizes,
        headChunkSizes: {},
        significance: defaultSizeChangesConfig.significance,
        budget: defaultSizeChangesConfig.budget,
      })
    ).toEqual([
      expect.objectContaining(negativeDelta),
      expect.objectContaining(negativeDelta),
      expect.objectContaining(negativeDelta),
    ]);
  });

  it('returns +100 delta when chunk is not in base but in head', () => {
    const positiveDelta: SizeChange = {
      currentSize: expect.any(Number),
      currentGzSize: expect.any(Number),
      currentBrSize: expect.any(Number),
      prevSize: null,
      prevGzSize: null,
      prevBrSize: null,
      sizeDiff: 100,
      gzSizeDiff: 100,
      brSizeDiff: 100,
      file: expect.any(String),
      significantFields: [
        SignificantField.sizeDiffBytes,
        SignificantField.sizeDiffPercent,
        SignificantField.gzSizeDiffPercent,
        SignificantField.brSizeDiffPercent,
      ],
    };

    expect(
      createDelta({
        baseChunkSizes: {},
        headChunkSizes,
        significance: defaultSizeChangesConfig.significance,
        budget: defaultSizeChangesConfig.budget,
      })
    ).toEqual([
      expect.objectContaining(positiveDelta),
      expect.objectContaining(positiveDelta),
      expect.objectContaining(positiveDelta),
    ]);
  });

  it('returns delta when chunk is in both base and head', () => {
    expect(
      createDelta({
        baseChunkSizes,
        headChunkSizes,
        significance: defaultSizeChangesConfig.significance,
        budget: defaultSizeChangesConfig.budget,
      })
    ).toMatchSnapshot();
  });

  it('orders largest changes first', () => {
    const base = {
      sgpTrainTimesPage: {
        css: {
          brSize: 5,
          gzSize: 7,
          size: 2,
        },
        js: {
          brSize: 1,
          gzSize: 2,
          size: 10,
        },
        mjs: {
          brSize: 3,
          gzSize: 4,
          size: 5,
        },
      },
    };
    const head = {
      sgpTrainTimesPage: {
        css: {
          brSize: 5,
          gzSize: 7,
          size: 2,
        },
        js: {
          brSize: 1,
          gzSize: 2,
          size: 15,
        },
        mjs: {
          brSize: 3,
          gzSize: 4,
          size: 4,
        },
      },
    };

    expect(
      createDelta({
        baseChunkSizes: base,
        headChunkSizes: head,
        significance: defaultSizeChangesConfig.significance,
        budget: defaultSizeChangesConfig.budget,
      })
    ).toEqual([
      expect.objectContaining({ file: 'sgpTrainTimesPage.js' }),
      expect.objectContaining({ file: 'sgpTrainTimesPage.mjs' }),
      expect.objectContaining({ file: 'sgpTrainTimesPage.css' }),
    ]);
  });

  it('orders alphabetically when diff value is same', () => {
    const baseSize = {
      brSize: 1,
      gzSize: 1,
      size: 1,
    };
    const base = {
      sgpTrainTimesPage: {
        css: baseSize,
        js: baseSize,
        mjs: baseSize,
      },
      euTrainTimesPage: {
        css: baseSize,
        js: baseSize,
        mjs: baseSize,
      },
    };

    expect(
      createDelta({
        baseChunkSizes: base,
        headChunkSizes: base,
        significance: defaultSizeChangesConfig.significance,
        budget: defaultSizeChangesConfig.budget,
      })
    ).toEqual([
      expect.objectContaining({ file: 'euTrainTimesPage.css' }),
      expect.objectContaining({ file: 'euTrainTimesPage.js' }),
      expect.objectContaining({ file: 'euTrainTimesPage.mjs' }),
      expect.objectContaining({ file: 'sgpTrainTimesPage.css' }),
      expect.objectContaining({ file: 'sgpTrainTimesPage.js' }),
      expect.objectContaining({ file: 'sgpTrainTimesPage.mjs' }),
    ]);
  });

  it('includes budget details', () => {
    const chunkName = 'sgpTrainTimesPage.js';
    const budgetConfig = { chunkName, size: '600KB' };
    const data = createDelta({
      baseChunkSizes,
      headChunkSizes,
      significance: defaultSizeChangesConfig.significance,
      budget: [budgetConfig],
    });
    const budgetData = data.find((d) => d.file === chunkName);

    expect(budgetData.budget).toEqual<BudgetResult>({
      amount: 90.04,
      matched: { chunkName: 'sgpTrainTimesPage.js', size: '600KB' },
      type: BudgetResultType.size,
    });
  });
});
