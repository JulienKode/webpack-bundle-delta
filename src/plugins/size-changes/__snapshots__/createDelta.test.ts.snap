// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDelta returns delta when chunk is in both base and head 1`] = `
Array [
  Object {
    "brSizeDiff": -74.2533822415611,
    "budget": null,
    "currentBrSize": 34674,
    "currentGzSize": 74867,
    "currentSize": 553190,
    "file": "sgpTrainTimesPage.js",
    "gzSizeDiff": -57.18631874510343,
    "prevBrSize": 134674,
    "prevGzSize": 174867,
    "prevSize": 653190,
    "significantFields": Array [
      "sizeDiffBytes",
      "sizeDiffPercent",
      "gzSizeDiffPercent",
      "brSizeDiffPercent",
    ],
    "sizeDiff": -15.309481161683431,
  },
  Object {
    "brSizeDiff": -2.700658960786432,
    "budget": null,
    "currentBrSize": 36028,
    "currentGzSize": 42562,
    "currentSize": 286997,
    "file": "sgpTrainTimesPage.css",
    "gzSizeDiff": -6.58443439708529,
    "prevBrSize": 37028,
    "prevGzSize": 45562,
    "prevSize": 296997,
    "significantFields": Array [
      "sizeDiffBytes",
      "gzSizeDiffPercent",
    ],
    "sizeDiff": -3.3670373774819273,
  },
  Object {
    "brSizeDiff": 0,
    "budget": null,
    "currentBrSize": 104674,
    "currentGzSize": 144867,
    "currentSize": 603190,
    "file": "sgpTrainTimesPage.mjs",
    "gzSizeDiff": 0,
    "prevBrSize": 104674,
    "prevGzSize": 144867,
    "prevSize": 603190,
    "significantFields": Array [],
    "sizeDiff": 0,
  },
]
`;
